### PR TITLE
Feat/oauth tv refresh flow

### PR DESCRIPTION
**Pull request by Pretzi**
## Description

[Monday ticket](https://applicaster.monday.com/boards/1615228456/pulses/3879804706)

`If merged this will add an observable operator to handle if multiple calls are made to the refreshToken function in case the token can only be used once, the token will now be only emitted to subscribers`

<details>
  <summary>Platforms Affected</summary>
  
- [ ] Android
- [ ] Android TV
- [ ] Apple TV (TVOS)
- [ ] Fire TV (FireOS)
- [ ] iOS
- [ ] LG TV (WebOS)
- [ ] Samsung TV (Tizen)
- [ ] Web
</details>

## Known or Pending Issues

`fill in this section with potential issues / limitations the reviewer(s) should know about`

## Screenshots

`If there is some kind of ui change please share screenshots`

## Version Info

`Does this update require a certain version of quickbrick, or does it rely on another pr to work?`

### CODE REVIEW

File: plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts [hunk 1]
```diff
@@ -71,6 +72,28 @@ export function isTokenExpired(expiresIn) {
...
 
+const authProvidersTokens = {};
+
+const handleRefreshToken = (dataModel): Observable<string> => {
+  const authProvider = dataModel?.authProvider;
+
+  if (!authProvidersTokens[authProvider]) {
+    authProvidersTokens[authProvider] = new Observable<string>((observable) => {
+      from(refreshToken(dataModel)).subscribe(
+        (res) => {
```
Comment (inckie) [inline plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts:83]: Remove from the authProvidersTokens map on completition

File: plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts [hunk 2]
```diff
@@ -104,14 +127,24 @@ export async function performRefreshToken(dataModel: DataModel) {
...
+    const refreshTokenObservable = handleRefreshToken(dataModel);
+
+    refreshTokenObservable.subscribe(
+      (refreshData) => {
+        log_debug(
+          "performRefreshToken: tokenObservable completed",
+          refreshData
+        );
+      },
+      (err) => {
```
Comment (inckie) [inline plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts:139]: throw is missing

File: plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts (hunk 1 above)
Comment (Pretzi) [reply plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts]: Done

File: plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts (hunk 2 above)
Comment (Pretzi) [reply plugins/quick-brick-login-oauth-tv-2.0/src/Components/Login/utils.ts]: Fixed

**Comment by Pretzi** [comment]
@inckie I did a refactor in the observable logic to convert the `handleRefreshToken` function into a promise and continue with the regular flow when the observable is complete.

**Comment by inckie** [comment]
Seems to work

```
2023-02-27 13:15:31.374 13055-13235 General                 com.bkfc                             D  Generalzapp_login_plugin_oauth_tv_2_0 performRefreshToken: before refresh
                                                                                                    
                                                                                                    	data: {refresh_token=b53e0ee3d19cb86f8d46d26cc200d6a425e188b97eb699af13273fcae4c3d9af, refreshSource={source=https://zapp-3085-bkfc.web.app/cleeng/refresh}, expired=true, refresh_date_unix_timestamp=1000.0}
2023-02-27 13:15:32.198 13055-13235 General                 com.bkfc                             D  Generalzapp_login_plugin_oauth_tv_2_0 performRefreshToken: before refresh
                                                                                                    
                                                                                                    	data: {refresh_token=b53e0ee3d19cb86f8d46d26cc200d6a425e188b97eb699af13273fcae4c3d9af, refreshSource={source=https://zapp-3085-bkfc.web.app/cleeng/refresh}, expired=true, refresh_date_unix_timestamp=1000.0}
2023-02-27 13:15:38.433 13055-13235 General                 com.bkfc                             D  Generalzapp_login_plugin_oauth_tv_2_0 performRefreshToken: refreshToken response
                                                                                                    
                                                                                                    	data: {access_token=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJjdXN0b21lcklkIjozMDU2NDEwMTEsInB1Ymxpc2hlcklkIjo4MTM1Mzk3OTgsImV4cCI6MTY3NzUyMjYzNn0.tvDL6tutDQjJwdTnuM-By14K53gX3NvrcA1gAT5BvcYlmV_dkTdUnCV9ACKLZ2Hw77sKkN7JF9c2W1pKckCkqVL17tnkb00_F-bELwqc4KE7hKvlpP9d_oDhLKpGR8xBUIFiDSNIm2gsM2zZDnq0WVM5teT12WvoKyGmyhZpj2vnfIPdfEPGWfn5BjHm-Vlk76VS42sFSsTpUQmhfJL6H7LrVCYXnFCzk1bM649Y4fqHRXswlRN9QMi8tE38jxp-RG-5NAytl891LciXOxU0h1PlfQi7ZEQnZLVamW3BTLt3r5k9CXcUrkXhgLlBICVhBIvR9MUxkRMo1XWsbTfPRg, refresh_token=e35206f80a6a1cff932880934e20572390ea9214244e708c68fd95646c0bbe91, expires_in=899.0, refresh_date_unix_timestamp=1.677522634E9}
2023-02-27 13:15:38.437 13055-13235 General                 com.bkfc                             D  Generalzapp_login_plugin_oauth_tv_2_0 performRefreshToken: tokenObservable completed
                                                                                                    
                                                                                                    	data: {data=null}

```

**Review by inckie — APPROVED** [review]

**Pull request merged**